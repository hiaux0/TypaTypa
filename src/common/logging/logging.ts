import { getCallerFunctionName } from "../modules/debugging";
import { CuLogger as Culogger, LogOptions } from "./localCulog";

export interface ShouldLogConfig {
  log?: boolean;
  maxLevel?: number; // 0 - don't log
  onlyLevels?: (undefined | number)[]; // only log specified levels
  logDepth?: number; // 1: 1-9, 2: 10-99, 3: 100-999
  allowedCallerNames?: string[];
  allowedCallerNameParts?: string[];
}

interface ILogOptions extends LogOptions {
  log?: boolean;
  focusedLogging?: boolean;

  measurePerf?: boolean;
  focusedPerf?: boolean;
  logPerf?: boolean;

  /**
   * Indicate to, that log can/should perform a reset
   */
  reset?: boolean;
  highlight?: boolean;
  shouldLogConfig?: ShouldLogConfig;
}

export const defaultShouldLogConfig: ShouldLogConfig = {
  log: true,
  maxLevel: 1,
  // onlyLevels: [1], // testing
  // onlyLevels: [3], // commands trace
  // onlyLevels: [4], // mergeKeybindingsV2 and overwriteAndAddExistingKeyBindingsV2
  // onlyLevels: [5], // context ids
  // onlyLevels: [,5], // trace potential commands
  // onlyLevels: [,5], // trace potential commands
  // onlyLevels: [,,1], // vim ui
  // onlyLevels: [,,2], // minimal command trace
  // onlyLevels: [,,3], // E2E v2
  // onlyLevels: [,,6], // All mode tracing
  logDepth: 1,
  allowedCallerNames: [],
  // allowedCallerNameParts: ["Vim"],
};

const DEFAULT_LOG_OPTIONS: ILogOptions = {
  log: false,
  focusedLogging: true,
  measurePerf: false,
  focusedPerf: true,
  logPerf: false,
  reset: false,
  shouldLogConfig: defaultShouldLogConfig,
};

export function logOptionsGuard(input: any): input is ILogOptions {
  if (typeof input !== "object") {
    return false;
  }

  const keys = Object.keys(input);
  const validKeys = [
    "log",
    "focusedLogging",
    "measurePerf",
    "focusedPerf",
    "logPerf",
    "reset",
    "highlight",
  ];
  for (const key of keys) {
    if (!validKeys.includes(key)) {
      return false;
    }
  }
  return true;
}

export class Logger {
  public readonly culogger: Culogger;

  constructor(
    public scope = "Aurelia",
    private readonly options: ILogOptions = DEFAULT_LOG_OPTIONS,
  ) {
    this.culogger = new Culogger({ scope });
    this.culogger.overwriteDefaultLogOtpions({
      logTable: true,
      logLevel: "INFO",
      focusedLogging: false,
      ...options,
      // logScope: false,
    });
  }

  log(...args: any[]) {
    let options = {} as ILogOptions;
    if (args.length > 2 && typeof args[args.length - 1] === "object") {
      options = args.pop() as ILogOptions;
    }
    const message = args.join(" ");

    /**
     * Wallaby logic.
     * Wallaby does not console.log from external library.
     */
    this.logMessage(message, options);
  }

  public shouldLog = (
    givenLevel?: number | (number | undefined)[],
    error?: Error,
  ) => {
    if (givenLevel === 0) return true;

    const callerName = getCallerFunctionName(error);
    const finalOptions = {
      ...defaultShouldLogConfig,
      ...this.options.shouldLogConfig,
    };
    const {
      maxLevel,
      logDepth,
      onlyLevels,
      allowedCallerNames,
      allowedCallerNameParts,
    } = finalOptions;
    // onlyLevels; /*?*/

    // Scope
    let scopeHasPart = "okay";
    if (allowedCallerNameParts) {
      const allowed = allowedCallerNameParts.find((namePart) =>
        this.scope.includes(namePart),
      );
      if (allowed) {
        scopeHasPart = allowed;
      }
    }

    // Level
    const asArray = Array.isArray(givenLevel) ? givenLevel : [givenLevel];
    const levelOkay = asArray.some((level, index) =>
      checkGivenLevel(level, index),
    );
    // levelOkay; /*?*/
    const nameOkay =
      allowedCallerNames?.find((name) => name.includes(callerName)) ?? true;
    // nameOkay; /*?*/
    const should = finalOptions.log && levelOkay && nameOkay && scopeHasPart;
    // should; /*?*/
    return !!should;

    function checkGivenLevel(
      level: number | undefined,
      index: number,
    ): boolean {
      // level; /*?*/
      if (!level) return false;
      let okay = false;
      if (level < 10 && logDepth === 1) {
        const mainLevel = level;
        if (onlyLevels && onlyLevels.length > 0) {
          okay = onlyLevels[index] === mainLevel;
        }
      } else if (level >= 10 && level < 100 && logDepth === 2) {
        const secondaryLevel = level % 10;
        // secondaryLevel; /*?*/
        if (onlyLevels && onlyLevels.length > 0) {
          okay = onlyLevels[index] === secondaryLevel;
          // okay; /*?*/
        }
      }
      return okay;
    }
  };

  private logMessage(
    message: string,
    options: ILogOptions = DEFAULT_LOG_OPTIONS,
  ) {
    const { log } = options;
    const loggedMessage = this.culogger.debug([message], {
      logLevel: "INFO",
      log,
      ...options,
    });
    if (loggedMessage.length > 0) {
      // console.log(loggedMessage[0]);
      if (loggedMessage.length > 1) {
        console.log("There are more log messages");
      }
    }
  }
}

export const shouldLog = new Logger("default").shouldLog;
// export function
