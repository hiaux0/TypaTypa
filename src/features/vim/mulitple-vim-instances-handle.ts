export function popVimInstanceId(): void {
  //if (window.activeVimInstancesIdMap.length > 1) {
  //  window.activeVimInstancesIdMap.pop();
  //}
  if (window.activeVimInstancesIdMap.length > 1) {
    // keep 1
    window.setTimeout(() => {
      window.activeVimInstancesIdMap.pop();
    }, 66);
  }
}
