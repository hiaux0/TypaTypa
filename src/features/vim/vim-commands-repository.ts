import { ModifiersType } from "../../common/modules/keybindings/app-keys";

export enum VIM_COMMAND {
  "backspace" = "backspace",
  "cancelAll" = "cancelAll",
  "clearLine" = "clearLine",
  "click" = "click",
  "createNewLine" = "createNewLine",
  "createNewLineAbove" = "createNewLineAbove",
  "copy" = "copy",
  "cursorRight" = "cursorRight",
  "cursorUp" = "cursorUp",
  "cursorLeft" = "cursorLeft",
  "cursorDown" = "cursorDown",
  "cursorWordForwardEnd" = "cursorWordForwardEnd",
  "cursorWordForwardStart" = "cursorWordForwardStart",
  "cursorBackwordsStartWord" = "cursorBackwordsStartWord",
  "cursorLineEnd" = "cursorLineEnd",
  "cursorLineStart" = "cursorLineStart",
  "delete" = "delete",
  "deleteInnerWord" = "deleteInnerWord",
  "deleteLine" = "deleteLine",
  "enterInsertAfterModeEnd" = "enterInsertAfterModeEnd",
  "enterInsertAfterMode" = "enterInsertAfterMode",
  "enterInsertMode" = "enterInsertMode",
  "enterInsertModeStart" = "enterInsertModeStart",
  "enterNormalMode" = "enterNormalMode",
  "enterVisualMode" = "enterVisualMode",
  "toggleFold" = "toggleFold",
  "goToFirstLine" = "goToFirstLine",
  "goToLastLine" = "goToLastLine",
  "indentRight" = "indentRight",
  "indentLeft" = "indentLeft",
  "joinLine" = "joinLine",
  "jumpNextBlock" = "jumpNextBlock",
  "jumpPreviousBlock" = "jumpPreviousBlock",
  "newLine" = "newLine",
  "paste" = "paste",
  "pasteVim" = "pasteVim",
  "redo" = "redo",
  "repeatLastCommand" = "repeatLastCommand",
  "replace" = "replace",
  "snippet" = "snippet",
  "space" = "space",
  "toCharacterAtBack" = "toCharacterAtBack",
  "toCharacterAt" = "toCharacterAt",
  "toCharacterAfterBack" = "toCharacterAfterBack",
  "toCharacterBefore" = "toCharacterBefore",
  "type" = "type",
  "undo" = "undo",
  "unfoldAll" = "unfoldAll",
  "yank" = "yank",
  "visualDelete" = "visualDelete",
  "visualInnerWord" = "visualInnerWord",
  "visualInnerBlock" = "visualInnerBlock",
  "visualStartLineWise" = "visualStartLineWise",
  "visualMoveToOtherEndOfMarkedArea" = "visualMoveToOtherEndOfMarkedArea",
  // custom
  "toggleCheckbox" = "toggleCheckbox",
  "hint" = "hint",
  "shift" = "shift",
  "customExecute" = "customExecute",
  "nothing" = "nothing",
}

export const VIM_MODE_COMMANDS = [
  VIM_COMMAND["enterInsertMode"],
  VIM_COMMAND["enterInsertModeStart"],
  VIM_COMMAND["enterNormalMode"],
  VIM_COMMAND["enterVisualMode"],
  VIM_COMMAND["visualStartLineWise"],
  VIM_COMMAND["createNewLine"],
  VIM_COMMAND["createNewLineAbove"],
  VIM_COMMAND["enterInsertAfterModeEnd"],
  VIM_COMMAND["enterInsertAfterMode"],
];

export const VIM_COMMANDS = [
  VIM_COMMAND.newLine,
  VIM_COMMAND["backspace"],
  VIM_COMMAND["createNewLine"],
  VIM_COMMAND["createNewLineAbove"],
  VIM_COMMAND["cursorRight"],
  VIM_COMMAND["cursorUp"],
  VIM_COMMAND["cursorLeft"],
  VIM_COMMAND["cursorDown"],
  VIM_COMMAND["cursorWordForwardEnd"],
  VIM_COMMAND["cursorWordForwardStart"],
  VIM_COMMAND["cursorBackwordsStartWord"],
  VIM_COMMAND["cursorLineEnd"],
  VIM_COMMAND["cursorLineStart"],
  VIM_COMMAND["delete"],
  VIM_COMMAND["deleteLine"],
  VIM_COMMAND["deleteInnerWord"],

  VIM_COMMAND["indentRight"],
  VIM_COMMAND["indentLeft"],
  VIM_COMMAND["joinLine"],
  VIM_COMMAND["jumpNextBlock"],
  VIM_COMMAND["jumpPreviousBlock"],
  VIM_COMMAND["newLine"],
  VIM_COMMAND["paste"],
  VIM_COMMAND["replace"],
  VIM_COMMAND["space"],
  VIM_COMMAND["toCharacterAtBack"],
  VIM_COMMAND["toCharacterAt"],
  VIM_COMMAND["toCharacterAfterBack"],
  VIM_COMMAND["toCharacterBefore"],
  VIM_COMMAND["type"],
  VIM_COMMAND["unfoldAll"],
  VIM_COMMAND["yank"],
  // visual
  VIM_COMMAND["visualDelete"],
  VIM_COMMAND["visualInnerWord"],
  VIM_COMMAND["visualInnerBlock"],
  VIM_COMMAND["visualStartLineWise"],
  VIM_COMMAND["visualMoveToOtherEndOfMarkedArea"],
  // ...VIM_MODE_COMMANDS,
  VIM_COMMAND["enterInsertMode"],
  VIM_COMMAND["enterInsertModeStart"],
  VIM_COMMAND["enterNormalMode"],
  VIM_COMMAND["enterVisualMode"],
];
export type VimCommandNames = keyof typeof VIM_COMMAND;

export const VIM_COMMANDS_THAT_CHANGE_TO_NORMAL_MODE = ["visualDelete"];

export interface VimCommand {
  /**
   * Modifiers: <mod>
   * Full: <modA><modB>keyAkeyB
   */
  key: string;
  command?: VimCommandNames;
  sequence?: string;
  execute?: () => void;
  args?: {
    text?: string;
  };
  desc?: string;
}

export interface SynonymKey {
  [key: string]: ModifiersType;
}
