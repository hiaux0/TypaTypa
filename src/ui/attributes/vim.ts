import { INode, ISignaler, resolve } from "aurelia";
import { VimStateClass } from "../../features/vim/vim-state";
import { IVimState, VimOptions } from "../../features/vim/vim-types";
import { VimInit } from "../../features/vim/VimInit";
import { IVimInputHandlerV2 } from "../../features/vim/VimInputHandlerV2";

export class VimCustomAttribute {
  private element = resolve(INode) as HTMLElement;
  vimState: IVimState;

  constructor(
    private vimInit: VimInit = resolve(VimInit),
    private vimInputHandler: IVimInputHandlerV2 = resolve(IVimInputHandlerV2),
  ) {}

  attached() {
    console.log(this.element);

    const vimState = VimStateClass.createEmpty();
    vimState.id = "bruh";
    vimState.lines = [{ text: "hi, world what are" }];
    const options: VimOptions = {
      vimState,
      vimId: vimState.id,
      container: this.element,
      childSelector: "vim-line",
      hooks: {
        vimStateUpdated: (vimState: IVimState) => {
          this.vimState = vimState;
        },
      },
    };
    this.vimInit.init(options);
    this.vimInputHandler.registerAndInit(options);
    this.vimState = vimState;
  }
}
