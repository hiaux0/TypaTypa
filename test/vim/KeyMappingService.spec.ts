import { describe, expect, test } from "vitest";
import { Logger } from "../../src/common/logging/logging";
import {
  KeyMappingService,
  overwriteExistingKeyBindingsV2,
} from "../../src/features/vim/vimCore/commands/KeyMappingService";
import { KeyBindingModes } from "../../src/features/vim/vim-types";
import { VimCommand } from "../../src/features/vim/vim-commands-repository";

const current: KeyBindingModes = {
  normal: [],
  NORMAL: [
    {
      key: "<Control>Enter",
    },
    {
      command: "cursorDown",
      desc: "Move to below cell when in INSERT mode",
      context: ["Grid"],
      key: "u",
    },
    {
      command: "cursorUp",
      desc: "Move to above cell when in INSERT mode",
      context: ["Grid"],
      key: "k",
    },
    {
      key: "<Escape>",
      desc: "Cancel edit and revert changes",
      command: "enterNormalMode",
    },
  ],
  insert: [],
  INSERT: [],
  visual: [],
  VISUAL: [
    {
      key: "y",
      desc: "yank",
      context: ["Grid"],
      command: "yank",
    },
  ],
  visualline: [],
  VISUALLINE: [
    {
      key: "d",
      command: "visualDelete",
    },
    {
      key: "o",
      command: "visualMoveToOtherEndOfMarkedArea",
    },
    {
      key: "x",
      command: "visualDelete",
    },
    {
      key: "<Control>c",
      command: "copy",
    },
    {
      key: "<Control>x",
      command: "cut",
    },
    {
      key: "<Control>z",
      command: "undo",
    },
    {
      key: "<Control><Shift>Z",
      command: "redo",
    },
    {
      key: "<Escape>",
      command: "enterNormalMode",
    },
    {
      key: "<ArrowLeft>",
      command: "cursorLeft",
    },
    {
      key: "<ArrowUp>",
      command: "cursorUp",
    },
    {
      key: "<ArrowRight>",
      command: "cursorRight",
    },
    {
      key: "<ArrowDown>",
      command: "cursorDown",
    },
    {
      key: "b",
      command: "cursorBackwordsStartWord",
    },
    {
      key: "e",
      command: "cursorWordForwardEnd",
    },
    {
      key: "gg",
      command: "goToFirstLine",
    },
    {
      key: "h",
      command: "cursorLeft",
    },
    {
      key: "k",
      command: "cursorUp",
    },
    {
      key: "l",
      command: "cursorRight",
    },
    {
      key: "u",
      command: "cursorDown",
    },
    {
      key: "w",
      command: "cursorWordForwardStart",
    },
    {
      key: "y",
      command: "yank",
    },
    {
      key: "<Shift>G",
      command: "goToLastLine",
    },
    {
      key: "<Shift>$",
      command: "cursorLineEnd",
    },
    {
      key: "$",
      command: "cursorLineEnd",
    },
    {
      key: "<Shift>^",
      command: "cursorLineStart",
    },
    {
      key: "^",
      command: "cursorLineStart",
    },
    {
      key: "<Shift>}",
      command: "jumpNextBlock",
    },
    {
      key: "}",
      command: "jumpNextBlock",
    },
    {
      key: "<Shift>{",
      command: "jumpPreviousBlock",
    },
    {
      key: "{",
      command: "jumpPreviousBlock",
    },
    {
      key: "<Shift>><Shift>>",
      command: "indentRight",
    },
    {
      key: "<Shift><<Shift><",
      command: "indentLeft",
    },
    {
      key: "<Shift>F",
      command: "toCharacterAtBack",
    },
    {
      key: "f",
      command: "toCharacterAt",
    },
    {
      key: "r",
      command: "replace",
    },
    {
      key: "T",
      command: "toCharacterAfterBack",
    },
    {
      key: "t",
      command: "toCharacterBefore",
    },
  ],
  CUSTOM: [],
  ALL: [
    {
      key: "<Space>cl",
      desc: "[cl]ear console",
      preventUndoRedo: true,
    },
  ],
  synonyms: {
    "<esc>": "<Escape>",
    escape: "<Escape>",
  },
};

const additional: KeyBindingModes = {
  normal: [],
  NORMAL: [
    {
      key: "<Space>tc",
      sequence: "^elrx",
    },
    {
      key: "<Space>fu",
      command: "unfoldAll",
    },
    {
      key: "a",
      command: "enterInsertAfterMode",
    },
    {
      key: "cc",
      command: "clearLine",
    },
    {
      key: "dd",
      command: "deleteLine",
    },
    {
      key: "diw",
      command: "deleteInnerWord",
    },
    {
      key: "i",
      command: "enterInsertMode",
      preventUndoRedo: true,
    },
    {
      key: "<Shift>I",
      command: "enterInsertModeStart",
    },
    {
      key: "J",
      command: "joinLine",
    },
    {
      key: "o",
      command: "createNewLine",
    },
    {
      key: "<Shift>O",
      command: "createNewLineAbove",
    },
    {
      key: "p",
      command: "pasteVim",
    },
    {
      key: "<Shift>P",
      command: "pasteVimBefore",
    },
    {
      key: "v",
      command: "enterVisualMode",
    },
    {
      key: "<Shift>V",
      command: "visualStartLineWise",
    },
    {
      key: "x",
      command: "delete",
    },
    {
      key: "za",
      command: "toggleFold",
    },
    {
      key: ".",
      command: "repeatLastCommand",
    },
    {
      key: "gh",
      command: "hint",
    },
    {
      key: "<Shift>A",
      command: "enterInsertAfterModeEnd",
    },
    {
      key: "<Control>v",
      command: "paste",
    },
    {
      key: "<Enter>",
      command: "newLine",
    },
    {
      key: "<Backspace>",
      command: "backspace",
    },
    {
      key: "<Meta>",
      command: "nothing",
    },
    {
      key: "<Control>c",
      command: "copy",
    },
    {
      key: "<Control>x",
      command: "cut",
    },
    {
      key: "<Control>z",
      command: "undo",
    },
    {
      key: "<Control><Shift>Z",
      command: "redo",
    },
    {
      key: "<Escape>",
      command: "enterNormalMode",
    },
    {
      key: "<ArrowLeft>",
      command: "cursorLeft",
    },
    {
      key: "<ArrowUp>",
      command: "cursorUp",
    },
    {
      key: "<ArrowRight>",
      command: "cursorRight",
    },
    {
      key: "<ArrowDown>",
      command: "cursorDown",
    },
    {
      key: "<Control>]",
      command: "indentRight",
    },
    {
      key: "<Control>[",
      command: "indentLeft",
    },
    {
      key: "b",
      command: "cursorBackwordsStartWord",
    },
    {
      key: "e",
      command: "cursorWordForwardEnd",
    },
    {
      key: "gg",
      command: "goToFirstLine",
    },
    {
      key: "h",
      command: "cursorLeft",
    },
    {
      key: "k",
      command: "cursorUp",
    },
    {
      key: "l",
      command: "cursorRight",
    },
    {
      key: "u",
      command: "cursorDown",
    },
    {
      key: "w",
      command: "cursorWordForwardStart",
    },
    {
      key: "y",
      command: "yank",
    },
    {
      key: "<Shift>G",
      command: "goToLastLine",
    },
    {
      key: "<Shift>$",
      command: "cursorLineEnd",
    },
    {
      key: "$",
      command: "cursorLineEnd",
    },
    {
      key: "<Shift>^",
      command: "cursorLineStart",
    },
    {
      key: "^",
      command: "cursorLineStart",
    },
    {
      key: "<Shift>}",
      command: "jumpNextBlock",
    },
    {
      key: "}",
      command: "jumpNextBlock",
    },
    {
      key: "<Shift>{",
      command: "jumpPreviousBlock",
    },
    {
      key: "{",
      command: "jumpPreviousBlock",
    },
    {
      key: "<Shift>><Shift>>",
      command: "indentRight",
    },
    {
      key: "<Shift><<Shift><",
      command: "indentLeft",
    },
    {
      key: "<Shift>F",
      command: "toCharacterAtBack",
    },
    {
      key: "f",
      command: "toCharacterAt",
    },
    {
      key: "r",
      command: "replace",
    },
    {
      key: "T",
      command: "toCharacterAfterBack",
    },
    {
      key: "t",
      command: "toCharacterBefore",
    },
  ],
  insert: [],
  INSERT: [
    {
      key: "<Backspace>",
      command: "backspace",
    },
    {
      key: "<Delete>",
      command: "delete",
    },
    {
      key: "<Enter>",
      command: "newLine",
    },
    {
      key: "<Shift>",
      command: "shift",
    },
    {
      key: "<Space>",
      command: "space",
    },
    {
      key: "<Control>",
      command: "nothing",
    },
    {
      key: "<Tab>",
      command: "indentRight",
    },
    {
      key: "<Shift><Tab>",
      command: "indentLeft",
    },
    {
      key: "<Control>c",
      command: "copy",
    },
    {
      key: "<Control>x",
      command: "cut",
    },
    {
      key: "<Control>z",
      command: "undo",
    },
    {
      key: "<Control><Shift>Z",
      command: "redo",
    },
    {
      key: "<Escape>",
      command: "enterNormalMode",
    },
    {
      key: "<ArrowLeft>",
      command: "cursorLeft",
    },
    {
      key: "<ArrowUp>",
      command: "cursorUp",
    },
    {
      key: "<ArrowRight>",
      command: "cursorRight",
    },
    {
      key: "<ArrowDown>",
      command: "cursorDown",
    },
  ],
  visual: [],
  VISUAL: [
    {
      key: "iw",
      command: "visualInnerWord",
    },
    {
      key: "ii",
      command: "visualInnerBlock",
    },
    {
      key: "d",
      command: "visualDelete",
    },
    {
      key: "o",
      command: "visualMoveToOtherEndOfMarkedArea",
    },
    {
      key: "x",
      command: "visualDelete",
    },
    {
      key: "<Control>c",
      command: "copy",
    },
    {
      key: "<Control>x",
      command: "cut",
    },
    {
      key: "<Control>z",
      command: "undo",
    },
    {
      key: "<Control><Shift>Z",
      command: "redo",
    },
    {
      key: "<Escape>",
      command: "enterNormalMode",
    },
    {
      key: "<ArrowLeft>",
      command: "cursorLeft",
    },
    {
      key: "<ArrowUp>",
      command: "cursorUp",
    },
    {
      key: "<ArrowRight>",
      command: "cursorRight",
    },
    {
      key: "<ArrowDown>",
      command: "cursorDown",
    },
    {
      key: "b",
      command: "cursorBackwordsStartWord",
    },
    {
      key: "e",
      command: "cursorWordForwardEnd",
    },
    {
      key: "gg",
      command: "goToFirstLine",
    },
    {
      key: "h",
      command: "cursorLeft",
    },
    {
      key: "k",
      command: "cursorUp",
    },
    {
      key: "l",
      command: "cursorRight",
    },
    {
      key: "u",
      command: "cursorDown",
    },
    {
      key: "w",
      command: "cursorWordForwardStart",
    },
    {
      key: "y",
      command: "yank",
    },
    {
      key: "<Shift>G",
      command: "goToLastLine",
    },
    {
      key: "<Shift>$",
      command: "cursorLineEnd",
    },
    {
      key: "$",
      command: "cursorLineEnd",
    },
    {
      key: "<Shift>^",
      command: "cursorLineStart",
    },
    {
      key: "^",
      command: "cursorLineStart",
    },
    {
      key: "<Shift>}",
      command: "jumpNextBlock",
    },
    {
      key: "}",
      command: "jumpNextBlock",
    },
    {
      key: "<Shift>{",
      command: "jumpPreviousBlock",
    },
    {
      key: "{",
      command: "jumpPreviousBlock",
    },
    {
      key: "<Shift>><Shift>>",
      command: "indentRight",
    },
    {
      key: "<Shift><<Shift><",
      command: "indentLeft",
    },
    {
      key: "<Shift>F",
      command: "toCharacterAtBack",
    },
    {
      key: "f",
      command: "toCharacterAt",
    },
    {
      key: "r",
      command: "replace",
    },
    {
      key: "T",
      command: "toCharacterAfterBack",
    },
    {
      key: "t",
      command: "toCharacterBefore",
    },
  ],
  visualline: [],
  VISUALLINE: [
    {
      key: "d",
      command: "visualDelete",
    },
    {
      key: "o",
      command: "visualMoveToOtherEndOfMarkedArea",
    },
    {
      key: "x",
      command: "visualDelete",
    },
    {
      key: "<Control>c",
      command: "copy",
    },
    {
      key: "<Control>x",
      command: "cut",
    },
    {
      key: "<Control>z",
      command: "undo",
    },
    {
      key: "<Control><Shift>Z",
      command: "redo",
    },
    {
      key: "<Escape>",
      command: "enterNormalMode",
    },
    {
      key: "<ArrowLeft>",
      command: "cursorLeft",
    },
    {
      key: "<ArrowUp>",
      command: "cursorUp",
    },
    {
      key: "<ArrowRight>",
      command: "cursorRight",
    },
    {
      key: "<ArrowDown>",
      command: "cursorDown",
    },
    {
      key: "b",
      command: "cursorBackwordsStartWord",
    },
    {
      key: "e",
      command: "cursorWordForwardEnd",
    },
    {
      key: "gg",
      command: "goToFirstLine",
    },
    {
      key: "h",
      command: "cursorLeft",
    },
    {
      key: "k",
      command: "cursorUp",
    },
    {
      key: "l",
      command: "cursorRight",
    },
    {
      key: "u",
      command: "cursorDown",
    },
    {
      key: "w",
      command: "cursorWordForwardStart",
    },
    {
      key: "y",
      command: "yank",
    },
    {
      key: "<Shift>G",
      command: "goToLastLine",
    },
    {
      key: "<Shift>$",
      command: "cursorLineEnd",
    },
    {
      key: "$",
      command: "cursorLineEnd",
    },
    {
      key: "<Shift>^",
      command: "cursorLineStart",
    },
    {
      key: "^",
      command: "cursorLineStart",
    },
    {
      key: "<Shift>}",
      command: "jumpNextBlock",
    },
    {
      key: "}",
      command: "jumpNextBlock",
    },
    {
      key: "<Shift>{",
      command: "jumpPreviousBlock",
    },
    {
      key: "{",
      command: "jumpPreviousBlock",
    },
    {
      key: "<Shift>><Shift>>",
      command: "indentRight",
    },
    {
      key: "<Shift><<Shift><",
      command: "indentLeft",
    },
    {
      key: "<Shift>F",
      command: "toCharacterAtBack",
    },
    {
      key: "f",
      command: "toCharacterAt",
    },
    {
      key: "r",
      command: "replace",
    },
    {
      key: "T",
      command: "toCharacterAfterBack",
    },
    {
      key: "t",
      command: "toCharacterBefore",
    },
  ],
  CUSTOM: [
    {
      key: "x",
      command: "delete",
    },
    {
      key: "<Control>c",
      command: "copy",
    },
    {
      key: "<Control>x",
      command: "cut",
    },
    {
      key: "<Control>z",
      command: "undo",
    },
    {
      key: "<Control><Shift>Z",
      command: "redo",
    },
    {
      key: "<Escape>",
      command: "enterNormalMode",
    },
    {
      key: "<ArrowLeft>",
      command: "cursorLeft",
    },
    {
      key: "<ArrowUp>",
      command: "cursorUp",
    },
    {
      key: "<ArrowRight>",
      command: "cursorRight",
    },
    {
      key: "<ArrowDown>",
      command: "cursorDown",
    },
    {
      key: "b",
      command: "cursorBackwordsStartWord",
    },
    {
      key: "e",
      command: "cursorWordForwardEnd",
    },
    {
      key: "gg",
      command: "goToFirstLine",
    },
    {
      key: "h",
      command: "cursorLeft",
    },
    {
      key: "k",
      command: "cursorUp",
    },
    {
      key: "l",
      command: "cursorRight",
    },
    {
      key: "u",
      command: "cursorDown",
    },
    {
      key: "w",
      command: "cursorWordForwardStart",
    },
    {
      key: "y",
      command: "yank",
    },
    {
      key: "<Shift>G",
      command: "goToLastLine",
    },
    {
      key: "<Shift>$",
      command: "cursorLineEnd",
    },
    {
      key: "$",
      command: "cursorLineEnd",
    },
    {
      key: "<Shift>^",
      command: "cursorLineStart",
    },
    {
      key: "^",
      command: "cursorLineStart",
    },
    {
      key: "<Shift>}",
      command: "jumpNextBlock",
    },
    {
      key: "}",
      command: "jumpNextBlock",
    },
    {
      key: "<Shift>{",
      command: "jumpPreviousBlock",
    },
    {
      key: "{",
      command: "jumpPreviousBlock",
    },
    {
      key: "<Shift>><Shift>>",
      command: "indentRight",
    },
    {
      key: "<Shift><<Shift><",
      command: "indentLeft",
    },
    {
      key: "<Shift>F",
      command: "toCharacterAtBack",
    },
    {
      key: "f",
      command: "toCharacterAt",
    },
    {
      key: "r",
      command: "replace",
    },
    {
      key: "T",
      command: "toCharacterAfterBack",
    },
    {
      key: "t",
      command: "toCharacterBefore",
    },
  ],
  ALL: [
    {
      key: "<Space>cl",
      desc: "[cl]ear console",
      preventUndoRedo: true,
    },
  ],
  synonyms: {
    "<esc>": "<Escape>",
    escape: "<Escape>",
  },
};

const logger = new Logger("VimInit.spec.ts", { terminalColor: "FgMagenta" });

describe("KeyMappingService", () => {
  test("mergeKeybindingsV2", () => {
    const keyMappingService = new KeyMappingService();
    const result = keyMappingService.mergeKeybindingsV2(current, additional);
    /*prettier-ignore*/ console.log("[KeyMappingService.spec.ts,996] result: ", result);
    expect(result).toBe(true);
  });

  test("overwriteExistingKeyBindingsV2", () => {
    const a = [
      {
        key: "<Enter>",
      },
    ];
    const b = [
      {
        key: "<Space>cl",
      },
    ];
    const result = overwriteExistingKeyBindingsV2(a, b);
    expect(result).toMatchSnapshot();
  });

  test("overwriteExistingKeyBindingsV2 - add, if not matching key", () => {
    const baseNormal: VimCommand[] = [
      {
        key: "k",
        command: "cursorUp",
      },
    ];

    const baseAll = [];

    const otherNormal: VimCommand[] = [
      {
        key: "<ArrowUp>",
        command: "cursorUp",
      },
    ];

    const otherAll = [];

    const result = overwriteExistingKeyBindingsV2(
      baseNormal,
      baseAll,
      otherNormal,
      otherAll,
    );
    expect(result).toMatchSnapshot();
  });

  test("overwriteExistingKeyBindingsV2 - add key to other", () => {
    const baseNormal: VimCommand[] = [
      {
        key: "k",
        command: "cursorUp",
      },
    ];

    const baseAll = [];

    const otherNormal: VimCommand[] = [
      {
        command: "cursorUp",
      },
    ];

    const otherAll = [];

    const result = overwriteExistingKeyBindingsV2(
      baseNormal,
      baseAll,
      otherNormal,
      otherAll,
    );
    expect(result).toMatchSnapshot();
  });

  test.only("overwriteExistingKeyBindingsV2 - add key to base", () => {
    const baseNormal: VimCommand[] = [
      {
        command: "cursorUp",
      },
    ];

    const baseAll = [];

    const otherNormal: VimCommand[] = [
      {
        key: "<ArrowUp>",
        command: "cursorUp",
      },
    ];

    const otherAll = [];

    const result = overwriteExistingKeyBindingsV2(
      baseNormal,
      baseAll,
      otherNormal,
      otherAll,
    );
    expect(result).toMatchSnapshot();
  });
});
