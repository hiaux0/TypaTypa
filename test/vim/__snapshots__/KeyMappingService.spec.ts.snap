// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KeyMappingService > overwriteExistingKeyBindingsV2 - Normal 1`] = `
[
  {
    "key": "<Enter>",
  },
  {
    "key": "<Space>cl",
  },
]
`;

exports[`KeyMappingService > overwriteExistingKeyBindingsV2 - add key to base 1`] = `
[
  {
    "command": "cursorUp",
  },
  {
    "command": "cursorUp",
    "key": "<ArrowUp>",
  },
]
`;

exports[`KeyMappingService > overwriteExistingKeyBindingsV2 - add key to other 1`] = `
[
  {
    "command": "cursorUp",
    "key": "k",
  },
]
`;

exports[`KeyMappingService > overwriteExistingKeyBindingsV2 1`] = `
[
  {
    "key": "<Enter>",
  },
  {
    "key": "<Space>cl",
  },
]
`;
